### PR TITLE
Feature/mvn packaging

### PR DESCRIPTION
**Pull request by PerilousApricot**
This isn't all the way there, but what I was talking about RE making this deployable via Maven

### CODE REVIEW

**Comment by mrow4a** [comment]
Perfect! We were looking for something like that, but did not have time to finish it! Thanks, I will test drive it. @mevangem @PerilousApricot

**Comment by mrow4a** [comment]
Did you manage to successfully run integration tests?  (Next step would probably to start CI)

File: pom.xml [hunk 1]
```diff
@@ -2,30 +2,80 @@
 	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
 	<modelVersion>4.0.0</modelVersion>
 	<groupId>ch.cern.eos</groupId>
-	<version>0.1.0</version>
-	<artifactId>hadoop-xrootd</artifactId>
-
+	<version>0.0.1-SNAPSHOT</version>
```
Comment (mrow4a) [inline pom.xml:5]: Please use 1.0.0-rc0 or 1.0.0-SNAPSHOT (both are fine with me)

File: pom.xml [hunk 2]
```diff
@@ -2,30 +2,80 @@
...
+	<artifactId>EOSConnector</artifactId>
```
Comment (mrow4a) [inline pom.xml:6]: Please don't change (https://github.com/mongodb/mongo-hadoop) - better use `hadoop-xrootd`

File: pom.xml [hunk 3]
```diff
@@ -2,30 +2,80 @@
...
+    </repositories>
+    <properties>
+        <maven.compiler.source>1.7</maven.compiler.source>
+        <maven.compiler.target>1.7</maven.compiler.target>
+    </properties>
 	<dependencies>
 		<dependency>
 			<groupId>org.apache.hadoop</groupId>
 			<artifactId>hadoop-common</artifactId>
-			<version>2.7.4</version>
```
Comment (mrow4a) [inline pom.xml:18]: This we would need to test. We are likely to look into 2.7.6

File: pom.xml [hunk 4]
```diff
@@ -2,30 +2,80 @@
...
+        </dependency>
+        <dependency>
+          <groupId>org.scijava</groupId>
+          <artifactId>native-lib-loader</artifactId>
+          <version>2.2.0</version>
+      </dependency>
+      <!--
+      <dependency>
+          <groupId>ch.cern.eos</groupId>
+          <artifactId>EOSConnector</artifactId>
```
Comment (mrow4a) [inline pom.xml:37]: hmm?

File: pom.xml [hunk 5]
```diff
@@ -2,30 +2,80 @@
...
+        </dependency>
+      -->
+      </dependencies>
 	<build>
-		<sourceDirectory>.</sourceDirectory>
-		<plugins>
+        <plugins>
+            <plugin>
+                <!-- Add JNI library -->
+                <groupId>com.github.maven-nar</groupId>
```
Comment (mrow4a) [inline pom.xml:48]: We would need to test it internally.

Comment (mrow4a) [reply pom.xml:48]: If this works, I buy you a beer :>

File: src/main/java/ch/cern/eos/XrootDBasedFileSystem.java [hunk 6]
```diff
@@ -191,9 +193,13 @@ private static void initJXrdCl() throws IOException {
 		try {
 			System.loadLibrary("jXrdCl");
 		} catch (UnsatisfiedLinkError e) {
-			e.printStackTrace();
-			eosDebugLogger.print("failed to load jXrdCl");
-			throw new IOException();
+            try {
+                NarSystem.loadLibrary();
```
Comment (mrow4a) [inline src/main/java/ch/cern/eos/XrootDBasedFileSystem.java:197]: ❤️

**Review by mrow4a — CHANGES_REQUESTED** [review]
Please look into comments. Additionaly, new commit with Documentation (including docker) adjustments.

Generally, I love this PR

File: pom.xml (hunk 1 above)
Comment (mrow4a) [reply pom.xml:5]: (this is first major release)

File: pom.xml [hunk 7]
```diff
@@ -2,30 +2,80 @@
...
+                  </linker>
+                  <libraries>
+                    <library>
+                      <type>jni</type>
+                      <narSystemPackage>ch.cern.eos.EOSConnector</narSystemPackage>
+                    </library>
+                  </libraries>
+                  <cpp>
+                    <systemIncludePaths>
+                      <systemIncludePath>/usr/include/xrootd</systemIncludePath>
```
Comment (mrow4a) [inline pom.xml:74]: hadoop-xrootd

File: pom.xml [hunk 8]
```diff
@@ -41,7 +91,7 @@
 						<descriptorRef>jar-with-dependencies</descriptorRef>
 					</descriptorRefs>
 				</configuration>
-			</plugin>
+            </plugin>
```
Comment (mrow4a) [inline pom.xml:94]: Typo?

File: src/main/java/ch/cern/eos/XrootDBasedFileSystem.java (hunk 6 above)
Comment (mrow4a) [reply src/main/java/ch/cern/eos/XrootDBasedFileSystem.java:197]: BUt please rethink it, try catch within try catch?

**Comment by mrow4a** [comment]
I will probably add here another commit, which uses maven in Dockerfile instead of make.

**Comment by PerilousApricot** [comment]
As you can tell, it still needs some work, but it gets a lot of the way there

File: pom.xml (hunk 4 above)
Comment (PerilousApricot) [reply pom.xml:37]: This is one of the issues I hit. I think that this needs to be split internally into two packages -- the NAR part with the shared libraries and the JAR part with the java code, then you can have the jar depend on the nar and they all get installed transitively

File: src/main/java/ch/cern/eos/XrootDBasedFileSystem.java (hunk 6 above)
Comment (PerilousApricot) [reply src/main/java/ch/cern/eos/XrootDBasedFileSystem.java:197]: I think that was just laziness. There's no real reason for it

File: pom.xml (hunk 3 above)
Comment (PerilousApricot) [reply pom.xml:18]: I think if/when it's moved to being built by CI, there will probably need to be builds against each major hadoop release and probably one "hadoop-provided" release that doesn't have any dependency

**Comment by PerilousApricot** [comment]
Looks like it should be possible to pass the x509 proxies around using the interface used to pass Kerberos tokens. Here since an alternate interface that looks good

https://github.com/zillow/aws-custom-credential-provider/blob/master/src/main/java/com/zillow/zda/data_lake/credential/RoleBasedAWSCredentialProvider.java

**Comment by mrow4a** [comment]
@PerilousApricot what is this META-INFO move you did? Is it related to what we were discussing? I will check this out

**Comment by mrow4a** [comment]
I am already rebased with your thing, so will test it all out on tuesday, and add the new commit

**Comment by PerilousApricot** [comment]
Check slack :)

**Comment by mrow4a** [comment]
Closing with https://github.com/cerndb/hadoop-xrootd/pull/9

**Pull request closed without merging**